### PR TITLE
Handling exception in getAssignment method

### PR DESCRIPTION
**Pull request by nisgoel-amazon**
Handling exception in getAssignment method

### Description
getAssignment method is used to assign shard replication task via persistent task cluster service. In case of stale entries present in persistent tasks in cluster state when this method is trying to assign task it is not handled properly.

Most of the times stale entries are not present in cluster state as we are clearing them at the time of cancelling the replication task but if cancelling is failed by any reason then these entries can remain in the cluster state.

 
### Issues Resolved
https://github.com/opensearch-project/cross-cluster-replication/issues/880
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by nisgoel-amazon** [comment]
Security and knn tests are failing due to known build failures with security plugin

```
ERROR][o.o.s.f.SecurityFilter   ] [leaderCluster-0] Unexpected exception java.lang.NoSuchMethodError: 'boolean org.opensearch.common.Strings.isNullOrEmpty(java.lang.String)'
»  java.lang.NoSuchMethodError: 'boolean org.opensearch.common.Strings.isNullOrEmpty(java.lang.String)'
»  	at org.opensearch.security.privileges.PrivilegesEvaluator.setUserInfoInThreadContext(PrivilegesEvaluator.java:204) ~[opensearch-security-3.0.0.0-SNAPSHOT.jar:3.0.0.0-SNAPSHOT]
»  	at org.opensearch.security.privileges.PrivilegesEvaluator.evaluate(PrivilegesEvaluator.java:248) ~[opensearch-security-3.0.0.0-SNAPSHOT.jar:3.0.0.0-SNAPSHOT]
»  	at org.opensearch.security.filter.SecurityFilter.apply0(SecurityFilter.java:303) [opensearch-security-3.0.0.0-SNAPSHOT.jar:3.0.0.0-SNAPSHOT]
»  	at org.opensearch.security.filter.SecurityFilter.apply(SecurityFilter.java:149) [opensearch-security-3.0.0.0-SNAPSHOT.jar:3.0.0.0-SNAPSHOT]
»  	at org.opensearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:216) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.action.support.TransportAction.execute(TransportAction.java:188) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.action.support.TransportAction.execute(TransportAction.java:107) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.client.node.NodeClient.executeLocally(NodeClient.java:110) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.client.node.NodeClient.doExecute(NodeClient.java:97) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.client.support.AbstractClient.execute(AbstractClient.java:476) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.client.support.AbstractClient$ClusterAdmin.execute(AbstractClient.java:788) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.client.support.AbstractClient$ClusterAdmin.health(AbstractClient.java:803) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.rest.action.admin.cluster.RestClusterHealthAction.lambda$prepareRequest$0(RestClusterHealthAction.java:85) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.rest.BaseRestHandler.handleRequest(BaseRestHandler.java:128) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.security.filter.SecurityRestFilter$1.handleRequest(SecurityRestFilter.java:127) [opensearch-security-3.0.0.0-SNAPSHOT.jar:3.0.0.0-SNAPSHOT]
»  	at org.opensearch.rest.RestController.dispatchRequest(RestController.java:320) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.rest.RestController.tryAllHandlers(RestController.java:411) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.rest.RestController.dispatchRequest(RestController.java:249) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
»  	at org.opensearch.security.ssl.http.netty.ValidatingDispatcher.dispatchRequest(ValidatingDispatcher.java:63) [opensearch-security-3.0.0.0-SNAPSHOT.jar:3.0.0.0-SNAPSHOT]
```

```
ERROR][o.o.b.OpenSearchUncaughtExceptionHandler] [followCluster-0] fatal error in thread [DefaultDispatcher-worker-1 @coroutine#1], exiting
»  java.lang.NoClassDefFoundError: org/opensearch/core/common/Strings
»  	at org.opensearch.commons.authuser.User.parse(User.java:162) ~[common-utils-3.0.0.0-SNAPSHOT.jar:?]
»  	at org.opensearch.replication.util.SecurityContext$Companion.fromSecurityThreadContext(SecurityContext.kt:65) ~[opensearch-cross-cluster-replication-3.0.0.0-SNAPSHOT.jar:3.0.0.0-SNAPSHOT]
»  	at org.opensearch.replication.action.index.TransportReplicateIndexAction$doExecute$1.invokeSuspend(TransportReplicateIndexAction.kt:66) ~[opensearch-cross-cluster-replication-3.0.0.0-SNAPSHOT.jar:3.0.0.0-SNAPSHOT]
```

**Comment by monu-aws** [comment]
@nisgoel-amazon will this be backported to all other branches ?

**Comment by nisgoel-amazon** [comment]
> @nisgoel-amazon will this be backported to all other branches ?

yes

File: src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationExecutor.kt [hunk 1]
```diff
@@ -56,9 +56,14 @@ class ShardReplicationExecutor(executor: String, private val clusterService : Cl
...
-        return Assignment(primaryShard.currentNodeId(), "node with primary shard")
+        try {
+            val primaryShard = clusterState.routingTable().shardRoutingTable(params.followerShardId).primaryShard()
+            if (!primaryShard.active()) return SHARD_NOT_ACTIVE
+            return Assignment(primaryShard.currentNodeId(), "node with primary shard")
+        } catch (e: Exception) {
+            log.error("Failed to assign shard replication task with id  ${params.followerShardId}", e)
+            return SHARD_NOT_ACTIVE
+        }
     }
```
Comment (saikaranam-amazon) [inline src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationExecutor.kt:67]: Let's add UTs for the assignment logic and can we validate the changes required for Index Replication task as well?

File: src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationExecutor.kt [hunk 2]
```diff
@@ -56,9 +56,14 @@ class ShardReplicationExecutor(executor: String, private val clusterService : Cl
...
 
     override fun getAssignment(params: ShardReplicationParams, clusterState: ClusterState) : Assignment {
-        val primaryShard = clusterState.routingTable().shardRoutingTable(params.followerShardId).primaryShard()
-        if (!primaryShard.active()) return SHARD_NOT_ACTIVE
-        return Assignment(primaryShard.currentNodeId(), "node with primary shard")
+        try {
+            val primaryShard = clusterState.routingTable().shardRoutingTable(params.followerShardId).primaryShard()
```
Comment (saikaranam-amazon) [inline src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationExecutor.kt:60]: For testing, will _stop API clear these entries when the shards are not active?

File: src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationExecutor.kt (hunk 1 above)
Comment (nisgoel-amazon) [reply src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationExecutor.kt:67]: Index Replication task does not attach to particular node, so there assignment can happen on any node.
Those tasks are working fine

File: src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationExecutor.kt (hunk 2 above)
Comment (nisgoel-amazon) [reply src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationExecutor.kt:60]: For this we are taking another issue
https://github.com/opensearch-project/cross-cluster-replication/issues/903

File: src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationExecutor.kt (hunk 1 above)
Comment (nisgoel-amazon) [reply src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationExecutor.kt:67]: UT has been added

**Review by ankitkala — APPROVED** [review]

**Review by saikaranam-amazon — APPROVED** [review]

**Pull request merged**